Server/scripts/fetchYearMaps.ts

```diff
@@ -88,15 +81,70 @@ const getBeatmapSet = memoizee(async (beatmap: APIBeatmap): Promise<Beatmapset>
         }
 
         // Check if new exists (remove if it does)
-        if (user.otherNames.some(v => v.name === beatmap.creator)) {
-            await user.otherNames.find(v => v.name === beatmap.creator)!.remove();
-            user.otherNames = user.otherNames.filter(v => v.name !== beatmap.creator);
+        if (user.otherNames.some(v => v.name === targetUser.username)) {
+            await user.otherNames.find(v => v.name === targetUser.username)!.remove();
+            user.otherNames = user.otherNames.filter(v => v.name !== targetUser.username);
         }
-        user.osu.username = beatmap.creator;
+        user.osu.username = targetUser.username;
         await user.save();
     }
+
+    return user;
+}, {
+    max: 200,
+    normalizer: ([user]) => `${user.username}`,
+});
+
+// Memoized method to create or fetch a BeatmapSet from DB.
+const existingSets: number[] = [];
+const getBeatmapSet = memoizee(async (beatmap: APIBeatmap): Promise<Beatmapset> => {
+    if(existingSets.includes(beatmap.setId))
+        return await Beatmapset.findOne(beatmap.setId) as Beatmapset;
+    let beatmapSet = new Beatmapset;
+    beatmapSet.ID = beatmap.setId;
+    beatmapSet.approvedDate = beatmap.approvedDate;
+    beatmapSet.submitDate = beatmap.submitDate;
+    beatmapSet.BPM = beatmap.bpm;
+    beatmapSet.artist = beatmap.artist;
+    beatmapSet.title = beatmap.title;
+    beatmapSet.genre = genres[beatmap.genre];
+    beatmapSet.language = langs[beatmap.language];
+    beatmapSet.tags = beatmap.tags.join(" ");
+    beatmapSet.favourites = beatmap.favoriteCount;
+
+    const user = await getUser({ username: beatmap.creator, userID: beatmap.creatorId });
     beatmapSet.creator = user;
 
+    // interOp call to pishi's BN site to get the user IDs of the BNs of a beatmapset.
+    // NOTE: The BN site only has nomination data from ~mid 2019 onward.
+    beatmapSet.rankers = [];
+    const beatmapEvents = (await axios.get(`https://bn.mappersguild.com/interOp/events/${beatmapSet.ID}`, {
+        headers:{
+            username: config.bn.username,
+            secret: config.bn.secret,
+        }
+    })).data as any[];
+    // Find the beatmap nominators for the map
+    if (beatmapEvents.length !== 0) {
+        const modeCount = beatmapEvents[0].modes.length;
+        const bns: number[] = [];
+        const nomEvents = beatmapEvents.filter(e => e.type === "nominate" || e.type === "qualify");
+        for (let i = nomEvents.length - 1; i >= 0; i--) {
+            if (!bns.some(bn => bn === nomEvents[i].userId))
+                bns.push(nomEvents[i].userId);
+            if (bns.length === modeCount * 2)
+                break;
```

Comment:
A set will also work here, saving you the need to check `!bns.some(bn => bn === nomEvents[i].userId)` for every event.

```suggestion
        const bns: Set<number> = new Set();
        const nomEvents = beatmapEvents.filter(e => e.type === "nominate" || e.type === "qualify");
        for (let i = nomEvents.length - 1; i >= 0; i--) {
            bns.add(nomEvents[i].userId);
            if (bns.size === modeCount * 2)
                break;
```